policy.go

```diff
@@ -0,0 +1,67 @@
+package proxyproto
+
+import "net"
+
+// PolicyFunc can be used to decide whether to trust the PROXY info from
+// upstream. If set, the connecting address is passed in as an argument.
+//
+// See below for the different policies.
+//
+// In case an error is returned the connection is denied.
+type PolicyFunc func(upstream net.Addr) (Policy, error)
```

Comment:
I think it would be good to add the type of incoming connection to the parameters. This allows you to change your response if the incoming is UNKNOWN, or to reject/accept v4/v6 without having to parse the address.

Reply:
Or alternately provide the actual proxy string as well as the net-derived Addr.

Reply:
I‘d go for passing the parsed header, not the proxy string. The decision about the policy has to be deferred to the first read then, and can‘t be done on accept. Which means any error from the actual ˋPolicyFuncˋ is only returned on read (or accessing one of the addresses, which is still incorrect as you pointed out below). Not sure if this is problematic from the user perspective, as it might be unexpected to receive such an error at that point after the user already received a connection?

Reply:
That's fair, it's nicer not to accept if you know you're not going to accept it. But perhaps then there needs to be a separate hook that allows you to control behavior based on the actual contents of the connection once it's accepted. v4/v6 is less of an issue but UNKNOWN is important -- if the proxy can't actually determine the originating address you may well want to drop the connection. (We added it to armon/go-proxyproto since a lot of users needed this behavior, so this is not a purely theoretical comment.)

Reply:
So what's the conclusion on this one?

Reply:
My current plan is to build the options variant mentioned below, and then see how passing a second hook for checking the proxy string looks like. It might be that I don‘t come up with a good design for that as I don’t have a usecase personally, but then  the API is still open to add this hook in another PR.

Reply:
Added an additional hook via 26a1d2a. In case the policy has been decided to `USE` the proxy header, a `Validator` can be used to inspect the parsed header. In case the validator returns an error this will mark the connection as erred and any subsequent read will return this error.